presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PayloadJoinOptimizer.java

```diff
@@ -183,18 +189,18 @@ public PlanNode visitJoin(JoinNode joinNode, RewriteContext<JoinContext> context
             ImmutableSet<VariableReferenceExpression> leftColumns = leftNode.getOutputVariables().stream().collect(toImmutableSet());
             Set<VariableReferenceExpression> joinKeys = extractJoinKeys(joinNode.getFilter(), joinNode.getCriteria());
 
-            if (!needsRewrite(joinNode.getType(), leftColumns, joinKeys)) {
+            ImmutableSet<VariableReferenceExpression> leftJoinKeys = intersection(joinKeys, leftColumns).immutableCopy();
```

Comment:
We usually just use Set<T> to represent, but always make sure they are ImmutableSet on the right side. You'll find this throughout the codebase, so let's be consistent.

Reply:
the function below (addKeys) expects an immutable set so it fails if I change it